crates/client/block_import/src/pre_validate.rs

```diff
@@ -0,0 +1,336 @@
+    // }
+
+    let compiled_class = contract_class
+        .compile()
+        .map_err(|e| BlockImportError::CompilationClassError { error: e.to_string(), class_hash })?;
+
+    let class_info = ClassInfo { contract_class, compiled_class_hash };
+
+    Ok(ConvertedClass { class_infos: (class_hash, class_info), class_compiled: (class_hash, compiled_class) })
+}
+
+fn transaction_hashes(
+    receipts: &[TransactionReceipt],
+    transactions: &[Transaction],
+    validation: &Validation,
+    block_n: Option<u64>,
+) -> Result<Vec<Felt>, BlockImportError> {
+    if receipts.len() != transactions.len() {
+        return Err(BlockImportError::TransactionEqualReceiptCount {
+            receipts: receipts.len(),
+            transactions: transactions.len(),
+        });
+    }
+
+    // mismatched block hash is allowed for blocks 1466..=2242 on mainnet
+    let is_special_trusted_case =
+        validation.chain_id == ChainId::Mainnet && block_n.is_some_and(|n| (1466..=2242).contains(&n));
+
+    if is_special_trusted_case || validation.trust_transaction_hashes {
+        Ok(receipts.iter().map(|r| r.transaction_hash()).collect())
+    } else {
+        transactions
+            .par_iter()
+            .enumerate()
+            .map(|(index, tx)| {
+                // Panic safety: receipt count was checked earlier
+                let got = receipts[index].transaction_hash();
+                let expected = tx.compute_hash(validation.chain_id.to_felt(), false, block_n);
+                if got != expected {
+                    return Err(BlockImportError::TransactionHash { index, got, expected });
+                }
+                Ok(got)
+            })
+            .collect()
+    }
+}
+
+/// Compute the transaction commitment for a block.
+fn transaction_commitment(block: &UnverifiedFullBlock, validation: &Validation) -> Result<Felt, BlockImportError> {
+    let starknet_version = block.header.protocol_version;
+
+    let transaction_hashes =
+        transaction_hashes(&block.receipts, &block.transactions, validation, block.unverified_block_number)?;
+
+    if let Some(expected) = block.commitments.transaction_count {
+        if expected != block.transactions.len() as u64 {
+            return Err(BlockImportError::TransactionCount { got: block.transactions.len() as _, expected });
+        }
+    }
+
+    // Compute transaction hashes
+    let tx_hashes_with_signature: Vec<_> = block
+        .transactions
+        .par_iter()
+        .zip(transaction_hashes)
+        .map(|(tx, tx_hash)| tx.compute_hash_with_signature(tx_hash, starknet_version))
+        .collect();
+
+    // Transaction commitment
+    let got = if starknet_version < StarknetVersion::STARKNET_VERSION_0_13_2 {
+        compute_merkle_root::<Pedersen>(&tx_hashes_with_signature)
+    } else {
+        compute_merkle_root::<Poseidon>(&tx_hashes_with_signature)
+    };
+
+    if let Some(expected) = block.commitments.transaction_commitment {
```

Comment:
this will only be called in full node mode right ? since in authority all commitments are set to None, maybe we can avoid that